File: 2020_year/12_month/3_week/jihun/algo/2667.py
import sys
sys.stdin = open("./dev/stdin")

N = int(input())

dy = [1, -1, 0, 0]
dx = [0, 0, 1, -1]

data = []
for _ in range(N):
    data.append(list(map(int, list(input()))))

checked = [[0] * (N) for _ in range(N)]

def dfs(y, x, idx):
    checked[y][x] = idx
    for i in range(4):
        yy = y + dy[i]
        xx = x + dx[i]
        if (0 <= yy < N and 0 <= xx < N and not checked[yy][xx] and data[yy][xx]):
            dfs(yy, xx, idx)

idx = 1
for y in range(N):
    for x in range(N):
        if (data[y][x] and not checked[y][x]):
            dfs(y, x, idx)
            idx += 1

print(idx-1)
answer = []
for i in range(idx-1):
    cnt = 0
    for row in checked:
        cnt += row.count(i+1)
    answer.append(cnt)
print('\n'.join(list(map(str, sorted(answer)))))

File: 2021_year/1_month/3_week/jihun/algo/2667.js
let fs = require('fs');
let input = fs.readFileSync('./dev/stdin').toString().split('\n')

const N = +input[0]

const data = []

for (let i = 1; i < N+1; i++) {
    data.push(input[i].split('').map(el => (+el)))
}

const checked = new Array(N).fill(new Array(N).fill(0))

const dy = [1, 0, 0, -1]
const dx = [0, 1, -1, 0]

const bfs = function(sy, sx, cnt) {
    const queue = []
    queue.push([sy, sx])
    checked[sy][sx] = cnt
    while (queue) {
        const temp  = queue.pop()
        const y = temp[0]
        const x = temp[1]
        for (let i = 0; i < 4; i++) {
            const ny = y + dy[i]
            const nx = x + dx[i]
            if (0 <= ny && ny < N && 0 <= nx && nx < N) {
                if (data[ny][nx] && !checked[ny][nx]) {
                    queue.push([ny, nx])
                    checked[ny][nx] = cnt
                }
            }
        }
    }
}

let cnt = 1
for (let y = 0; y < N; y++) {
    for (let x = 0; x < N; x++) {
        if (data[y][x] && !checked[y][x]) {
            bfs(y, x, cnt)
        }
    }
}